src/traces/heatmap/clean_2d_array.js

```diff
@@ -30,12 +30,23 @@ module.exports = function clean2dArray(zOld, transpose) {
         old2new = function(zOld, i, j) { return zOld[i][j]; };
     }
 
+    var xMap = function(i) {return i;};
+    var yMap = function(i) {return i;};
+    if(trace && trace.type === 'heatmap') {
```

Comment:
Hmm. Why _just_ `heatmap`?

Reply:
This was the low-hanging fruit. I guess I should enlarge the scope of this PR and implement sorting of categorical axes for `contour`, `histogram2d` and all other traces relying on `heatmap`'s calc routine.

Reply:
Yes please!

Reply:
definitely useful for `histogram2d`! I have qualms from a viz-theory standpoint about sorting an axis of a contour plot: if you're drawing contours, which interpolate between data points, it strongly implies a continuous axis, so you really have no business with categories unless they're some sort of a hack of what's really a continuous axis... and sorting a continuous axis is meaningless.

That said, we allow it in the schema, therefore it's a bug if it doesn't work!

Reply:
Histogram and contour are now working in 6b06516

By allowing all traces (except `carpet` and `contourcarpet`), the only test that would fail was in `bar_test.js`. In it, we would create a `contour` trace without `x` and `y` data. In that case, `trace._x` and `trace._y` would be empty and it failed. To fix this, we check whether those arrays are empty and ignore categorical sorting if they are. After all, if the data has no categories attached to it, it's not clear how it should be sorted.